xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,398 @@
+			}
+		}
+
+		gsb.balancerCurrent = gsb.balancerPending
+		gsb.balancerPending = nil
+	}()
+}
+
+// Helper function that checks if the balancer passed in is current or pending.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) balancerCurrentOrPending(bal balancer.Balancer) bool {
+	return bal == gsb.balancerCurrent || bal == gsb.balancerPending
+}
+
+func (gsb *gracefulSwitchBalancer) newSubConn(bal balancer.Balancer, addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	if gsb.closed.HasFired() {
+		// logger?
+		return nil, errBalancerClosed
+	}
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return nil, fmt.Errorf("%T at address %p that called NewSubConn is deleted", bal, bal)
+	}
+	sc, err := gsb.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	gsb.scToSubBalancer[sc] = bal
+	return sc, nil
+}
+
+// close guard for the three functions below?
+func (gsb *gracefulSwitchBalancer) resolveNow(bal balancer.Balancer, opts resolver.ResolveNowOptions) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	// If the resolver specifies an update with a whole new type of policy,
+	// there is no reason to forward a ResolveNow call from the balancer being
+	// switched from, as the most recent update from the Resolver does not
+	// concern the balancer being switched from.
+	if bal == gsb.balancerCurrent && gsb.balancerPending != nil {
+		return
+	}
+	gsb.cc.ResolveNow(opts)
+}
+
+func (gsb *gracefulSwitchBalancer) removeSubConn(bal balancer.Balancer, sc balancer.SubConn) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+	delete(gsb.scToSubBalancer, sc)
+	gsb.cc.RemoveSubConn(sc)
+}
+
+func (gsb *gracefulSwitchBalancer) updateAddresses(bal balancer.Balancer, sc balancer.SubConn, addrs []resolver.Address) {
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+	gsb.cc.UpdateAddresses(sc, addrs)
+}
+
+// SwitchTo gracefully switches to the new balancer. This function must be
+// called synchronously alongside the rest of the balancer.Balancer methods this
+// Graceful Switch Balancer implements.
+func (gsb *gracefulSwitchBalancer) SwitchTo(builder balancer.Builder) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+	gsb.mu.Lock()
+	ccw := &clientConnWrapper{
+		ClientConn: gsb.cc,
+		gsb:        gsb,
+	}
+	newBalancer := builder.Build(ccw, gsb.bOpts)
+	if newBalancer == nil {
+		// Can this even ever happen?
+		return balancer.ErrBadResolverState
+	}
+	ccw.bal = newBalancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = newBalancer
+		gsb.mu.Unlock()
+		return nil
+	}
+	var balToClose balancer.Balancer
+	gsb.pendingState = balancer.State{
+		ConnectivityState: connectivity.Connecting,
+		Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+	}
```

Comment:
Move this up above the previous `if` and you don't need to initialize this in `newGracefulSwitchBalancer` do you?